tests/test_models.py

```diff
@@ -149,31 +153,126 @@ def test_serialize_a_customer(self):
-    
+        """It should Deserialize a Customer and its Addresses"""
+        customer = CustomerFactory()
+        customer.addresses.append(AddressFactory())
+        customer.create()
+        serial_customer = customer.serialize()
+        new_customer = Customer()
+        new_customer.deserialize(serial_customer)
+        self.assertEqual(new_customer.id, None)
+        self.assertEqual(new_customer.first_name, customer.first_name)
+        self.assertEqual(new_customer.last_name, customer.last_name)
+        self.assertEqual(new_customer.email, customer.email)
+        self.assertEqual(new_customer.password, customer.password)
+
     def test_deserialize_missing_data(self):
         """It should not deserialize a Customer with missing data"""
         data = {"id": 1, "first_name": "Kitty", "email": "kitty@gmail.com"}
         customer = Customer()
         self.assertRaises(DataValidationError, customer.deserialize, data)
-    
+
     def test_deserialize_bad_data(self):
         """It should not deserialize bad data"""
         data = "this is not a dictionary"
         customer = Customer()
         self.assertRaises(DataValidationError, customer.deserialize, data)
 
+    def test_deserialize_address_key_error(self):
+        """It should not Deserialize an address with a KeyError"""
+        address = Address()
+        self.assertRaises(DataValidationError, address.deserialize, {})
+
+    def test_deserialize_address_type_error(self):
+        """It should not Deserialize an address with a TypeError"""
+        address = Address()
+        self.assertRaises(DataValidationError, address.deserialize, [])
+
+    def test_add_customer_address(self):
+        """It should Create an Customer with an Address and add it to the database"""
+        customers = Customer.all()
+        self.assertEqual(customers, [])
+        customer = CustomerFactory()
+        address = AddressFactory(customer=customer)
+        customer.addresses.append(address)
+        customer.create()
+        # Assert that it was assigned an id and shows up in the database
+        self.assertIsNotNone(customer.id)
+        customers = Customer.all()
+        self.assertEqual(len(customers), 1)
+
+        new_customer = customer.find(customer.id)
+        self.assertEqual(new_customer.addresses[0].street, address.street)
+
+        address2 = AddressFactory(customer=customer)
+        customer.addresses.append(address2)
+        customer.update()
+
+        new_customer = Customer.find(customer.id)
+        self.assertEqual(len(new_customer.addresses), 2)
+        self.assertEqual(new_customer.addresses[1].street, address2.street)
+
+    def test_update_account_address(self):
+        """It should Update an Customer's address"""
+        customers = Customer.all()
+        self.assertEqual(customers, [])
+
+        customer = CustomerFactory()
+        address = AddressFactory(customer=customer)
+        customer.create()
```

Comment:
address.create() missing

Reply:
I am not sure if we really need address.create() following the repo shared by the Prof for the sample account [repo](https://github.com/nyu-devops/sample-accounts/blob/master/tests/test_models.py).

Reply:
update() function added in the Customer class.